doc/development/06_ca_manager.md

```diff
@@ -0,0 +1,976 @@
+check with a parent for resource entitlements (default: every 10 mins) would
+result in an excessive amount of commands and events in the history. In itself
+that is not an issue as long as these historical entries would be valuable
+to rebuilding or understanding the current state of a CA, but the **automatic**
+generation of CRLs and manifests, and historical connections between a CA and
+its parent or repository are not very interesting in this regard.
+
+As a result the `CaManager` is fairly complex because it needs to handle
+these three components in parallel, together with a `CaLocks` structure
+which is used lock read and write access to individual CAs that it manages.
+
+And, for completeness, this component has access to a `KrillSigner` that
+can be used to sign things, and the central `Config` that contains runtime
+settings such as the frequency by which to republish new manifests and CRLs.
+
+Combined we then get this structure:
+
+```rust
+#[derive(Clone)]
+pub struct CaManager {
+    ca_store: Arc<AggregateStore<CertAuth>>,
+    ca_objects_store: Arc<CaObjectsStore>,
+    status_store: Arc<Mutex<StatusStore>>,
+    locks: Arc<CaLocks>,
+    config: Arc<Config>,
+    signer: Arc<KrillSigner>,
+}
+```
+
+Initialization
+--------------
+
+The `CaManager` is instantiated when Krill starts. All its components ultimately
+rely on the `KeyValueStore` - which currently only supports a disk based back-end,
+but which can be modified to support other storage options in future.
+
+Let's first have a look at the initialization code and the included comments, and
+then explain a bit more below.
+
+```rust
+impl CaManager {
+    /// Builds a new CaServer. Will return an error if the CA store cannot be initialized.
+    pub async fn build(config: Arc<Config>, mq: Arc<MessageQueue>, signer: Arc<KrillSigner>) -> KrillResult<Self> {
+        // Create the AggregateStore for the event-sourced `CertAuth` structures that handle
+        // most CA functions.
+        let mut ca_store = AggregateStore::<CertAuth>::disk(&config.data_dir, CASERVER_DIR)?;
+
+        if config.always_recover_data {
+            // If the user chose to 'always recover data' then do so.
+            // This is slow, but it will ensure that all commands and events are accounted for,
+            // and there are no incomplete changes where some but not all files for a change were
+            // written to disk.
+            ca_store.recover()?;
+        } else if let Err(e) = ca_store.warm() {
+            // Otherwise we just tried to 'warm' the cache. This serves two purposes:
+            // 1. this ensures that all `CertAuth` structs are available in memory
+            // 2. this ensures that there are no apparent data issues
+            //
+            // If there are issues, then complain and try to recover.
+            error!(
+                "Could not warm up cache, data seems corrupt. Will try to recover!! Error was: {}",
```

Comment:
Could out-of-memory also be the cause of this failure?

Reply:
I think this would cause a panic, and Krill would exit (as configured in the Cargo.toml). But, I believe it would not be likely that this will result in excessive memory usage.